### PR TITLE
Use different conv layout optimization heuristics for inference

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114600

While many models regress in training when converted to channels last, in inference the results are quite different. Almost all of the models experienced a speedup when converted to channels last. There were a few big regressions in torchbench - `timm_regnet` from `1.4343 → 1.0573` and `timm_resnet` from `1.7484 → 1.2868`.

 I used a modified script of the operator benchmarks [here](https://gist.github.com/eellison/e11dc645412f52e8b45fb26ba6f9f6a1) to measure the average speedup of convolutions across all of the input shapes found in torchbench according to the existing classifications that @shunting314 used - grouped convs, small channel convs, convolution with larger in-channel than out-channel. Only grouped convolutions benchmarked as a slowdown in inference. 

I updated the inference heuristic to multiply the flops of each conv with its predicted speedup/slowdown in channels last. With this heuristic the two previously regressing models no longer regress. 

Speeds up inference for torchbench ~8% and timm ~6%. The motivating model here was SDXL which now hits channels last and improves 10%.

There were some models that were sped up in training when forcing channels last (along with a number of regressions). It's possible there is some speedup in training to be had with additional heuristics. We could also have more granular classification/predictions which might benefit both training and inference.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -3043,8 +3084,8 @@ def forward(arg6, arg7, arg16):
             # Mismatched elements: 127 / 746496 (0.0%)
             # Greatest absolute difference: 0.0009765625 at index (1, 62, 7, 16) (up to 1e-05 allowed)
             # Greatest relative difference: 0.05187467899332306 at index (14, 18, 11, 0) (up to 0.001 allowed)
-            atol=1e-3,
-            rtol=0.001,
+            atol=3e-3,
+            rtol=2,
```
Comment (jansel) [inline test/inductor/test_torchinductor.py:3088]: rtol=2 seems pretty bad... what is going on here?

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 2]
```diff
@@ -51,6 +51,7 @@
     "test_kwargs_dynamic_shapes": TestFailure(("cpu",)),
     # calling div on only symint args
     "test_AllenaiLongformerBase_repro_dynamic_shapes": TestFailure(("cpu", "cuda")),
+    "test_conv_inference_heuristics_dynamic_shapes": TestFailure("cuda"),
```
Comment (jansel) [inline test/inductor/test_torchinductor_dynamic_shapes.py:54]: why?

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -296,6 +301,73 @@ def decide_layout_opt(gm) -> bool:
...
+
+                    flop_counts[node_type] += counted_flops
+                else:
+                    log.debug("Conv inputs meta not found")
+
+            # average benchmarked speedup / slowdown, < 1 is speedup
+            GROUPED_MULTIPLIER = 1.358
+            DEFAULT_MULTIPLIER = 0.823
+            IN_OUT_MULTIPLIER = 0.725
+            SMALL_MULTIPLIER = 0.783
```
Comment (jansel) [inline torch/_inductor/graph.py:352]: Comment about how there we generated?  How to update?

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (eellison) [reply test/inductor/test_torchinductor.py:3088]: Sometimes channels last kernel can produce very different numerics especially with random input data. I'll see if I can repro this with eager channels last permutation.

Comment (eellison) [reply test/inductor/test_torchinductor.py:3088]: You get similarly high different rtol just by comparing eager channels last and eager contiguous, and with just the convolution:

> Mismatched elements: 3084650 / 3097600 (99.6%)
Greatest absolute difference: 0.02942657470703125 at index (7, 19, 34, 22) (up to 1e-05 allowed)
Greatest relative difference: 361.1468811035156 at index (3, 20, 48, 11) (up to 1.3e-06 allowed)

**Review by jansel — APPROVED** [review]

**Review by shunting314 — APPROVED** [review]

**Comment by eellison** [comment]
@shunting314 the speedup is for inference relative to main.

**Comment by stncil** [comment]
Is there any reason as to why this speedup/slowdown occurs and also why the behavior changes during training and inference>

**Pull request closed without merging**